### PR TITLE
Cnn transformer

### LINKED ISSUE #2: Implement and integrate backbone Architecture: pure CNN

### LINKED ISSUE #3: Implement and integrate backbone Architecture: full transformer

### PR DESCRIPTION
**Pull request by Ahwodafo**
Created artificial t show changes, since I already pushed onto main

### CODE REVIEW

File: models/cnn/cnn.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class CNN(nn.Module):
+
```
Comment (RaphaelKreft) [inline models/cnn/cnn.py:17]: Please add a comment/docstring explaining from where this architecture stems (ex. paper reference) and quickly explain the structure.

Also explain what the different parameters do and add a demo config that one can put into the json file (if your model supports configuration as so)

File: models/cnn/cnn.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        self.pretrain = pretrain
+        # architecture
+        self.init_layer = self.make_layers(in_channels=1, out_channels=64, n_layers=2, maxpool_size=None, first=True)
+        self.layer1 = self.make_layers(in_channels=64, out_channels=128, n_layers=2, maxpool_size=5)
+        self.layer2 = self.make_layers(in_channels=128, out_channels=192, n_layers=3, maxpool_size=5)
+        self.layer3 = self.make_layers(in_channels=192, out_channels=256, n_layers=3, maxpool_size=5)
+        self.layer4 = self.make_layers(in_channels=256, out_channels=256, n_layers=3, maxpool_size=5)
+            
+        self.fp_dim = 128
```
Comment (RaphaelKreft) [inline models/cnn/cnn.py:27]: Please use more expressive name: I guess "fp" = "feature_pyramid"?

File: models/cnn/cnn.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        for m in self.modules():
+            if isinstance(m, nn.Conv1d):
+                nn.init.kaiming_normal_(m.weight, mode='fan_out', nonlinearity='relu')
+                if m.bias is not None:
+                    nn.init.constant_(m.bias, 0)
+            elif isinstance(m, nn.BatchNorm1d):
+                nn.init.constant_(m.weight, 1)
+                nn.init.constant_(m.bias, 0)
+
+    def make_layers(self, in_channels, out_channels, n_layers, maxpool_size, first=False):
```
Comment (RaphaelKreft) [inline models/cnn/cnn.py]: Please add type hints and add a quick docstrinf/comment what this method does exactly

File: models/cnn/cnn.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+            conv1d = nn.Conv1d(in_channels, out_channels, kernel_size=3, padding=1)
+            layers += [conv1d, nn.BatchNorm1d(out_channels)]
+            if i == n_layers - 1:
+                layers += [ChannelGate(in_channels)]
+            layers += [nn.PReLU()]
+            in_channels = out_channels
+
+        return nn.Sequential(*layers)
+
+    def forward(self, x):
```
Comment (RaphaelKreft) [inline models/cnn/cnn.py:50]: Please add type hints  and docstring/comment what this method outputs/does depending on the configuration (ex. how the self.pretrain flag is set)

File: models/cnn/cnn_backbone.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
+import torch.nn as nn
+import torch.nn.functional as F
+
+from .cnn import CNN
+
+class CNN(nn.Module):
```
Comment (RaphaelKreft) [inline models/cnn/cnn_backbone.py]: What is this backbone exactly for? Please add a doc/comment and please explain how this is going to be integrated into the pipeline.

Comment (RaphaelKreft) [reply models/cnn/cnn_backbone.py]: I see its partly explained in the model readme. But I actually think that the backbone models are not really needed as the backbones are used in the main_model file.

What would be good is to add the option tho the DL model to have a decoder activated or not. This is needed for  Masked prediction where every model that should be applicable to MP must have an autoencoder mode and an encoder only mode.

**Review by RaphaelKreft — CHANGES_REQUESTED** [review]
In general good work!

### Some things to change:

- Add comments and docstrings to your implementation, especially on class level to explain where the architecture comes from and what the parameters/configs are (see my comments in cnn.py for more details)
- Add type hints where possible. This makes debugging/underdtanding easier later on

### Major Integration Challenges we need to discuss

- We need an abstract class thats having a clear way to access the model and a common way of configuration
- THis will allow to specify any kind of model and specify for each which training modes (CL, MP) are supported

File: models/cnn/cnn.py (hunk 1 above)
Comment (Ahwodafo) [reply models/cnn/cnn.py:17]: Added the docstring. While I do not have config support, they can be added if necessary. As of right now, I have added everything as default variables except for the pretrain flag, which (in Theory) can be removed

File: models/cnn/cnn.py (hunk 2 above)
Comment (Ahwodafo) [reply models/cnn/cnn.py:27]: 👍

File: models/cnn/cnn_backbone.py (hunk 5 above)
Comment (Ahwodafo) [reply models/cnn/cnn_backbone.py]: The backbone files are needed, because different things have to be called depending on the model. For example the transformer requires the positional embeddings, the CNN (if we are going to replicate it as the paper does) uses a mlp after the CNN for the CRL. 

However, the backbone files are still rough drafts and can be changed/removed if we aim to do different things with the model

**Comment by Ahwodafo** [comment]
Quick thing to note. The decoder is finicky, because the original model does some weird padding sometimes. It works for input size 3k (or any other input size that is divisible by 500), but the maxpool_size needs to be adjusted if that is not the case

File: models/cnn/cnn_backbone.py (hunk 5 above)
Comment (RaphaelKreft) [reply models/cnn/cnn_backbone.py]: Alright, lets keep them for now. We can change as needed

File: models/cnn/cnn.py (hunk 1 above)
Comment (RaphaelKreft) [reply models/cnn/cnn.py:17]: I will add the config options in my PR, no worries. The pre-train flag can be removed I think but I will check and leave it as necessary

**Review by RaphaelKreft — APPROVED** [review]
LGTM for me now. There are still open questions to resolve and there may be changes to the structure of the models. But this can be dealt with in respective PR's.
